### PR TITLE
Rustup to rustc 1.39.0-nightly (acf7b50c7 2019-09-25) 

### PR DESCRIPTION
**Pull request by Manishearth**
changelog: none

fixes https://github.com/rust-lang/rust/issues/64777

r? @phansch @yaahc

### CODE REVIEW

**Comment by Manishearth** [comment]
Working on fixing the or-patterns breakage.

File: clippy_lints/src/cognitive_complexity.rs [hunk 1]
```diff
@@ -112,8 +112,7 @@ impl<'tcx> Visitor<'tcx> for CCHelper {
         walk_expr(self, e);
         match e.node {
             ExprKind::Match(_, ref arms, _) => {
-                let arms_n: u64 = arms.iter().map(|arm| arm.pats.len() as u64).sum();
-                if arms_n > 1 {
+                if arms.len() > 1 {
```
Comment (Centril) [inline clippy_lints/src/cognitive_complexity.rs]: If I'm reading the intent of this code you might want to recurse and count each or-pattern as `cc += 1` but I think the change you made makes more sense.

File: clippy_lints/src/copies.rs [hunk 2]
```diff
@@ -213,27 +213,22 @@ fn lint_match_arms<'tcx>(cx: &LateContext<'_, 'tcx>, expr: &Expr) {
...
+                        // hiding all the subsequent arms, and rust won't compile
+                        db.span_note(
+                            i.body.span,
+                            &format!(
+                                "`{}` has the same arm body as the `_` wildcard, consider removing it`",
+                                lhs
+                            ),
+                        );
+                    } else {
+                        db.span_help(i.pat.span, &format!("consider refactoring into `{} | {}`", lhs, rhs));
```
Comment (Centril) [inline clippy_lints/src/copies.rs]: Aside: In the spirit of this lint, in the future you may want a lint that suggests `C(p0) | C(p1)` => `C(p0 | p1)`.

File: clippy_lints/src/matches.rs [hunk 3]
```diff
@@ -605,12 +598,7 @@ fn check_match_ref_pats(cx: &LateContext<'_, '_>, ex: &Expr, arms: &[Arm], expr:
 }
 
 fn check_match_as_ref(cx: &LateContext<'_, '_>, ex: &Expr, arms: &[Arm], expr: &Expr) {
-    if arms.len() == 2
-        && arms[0].pats.len() == 1
-        && arms[0].guard.is_none()
-        && arms[1].pats.len() == 1
-        && arms[1].guard.is_none()
-    {
+    if arms.len() == 2 && arms[0].guard.is_none() && arms[1].guard.is_none() {
```
Comment (Centril) [inline clippy_lints/src/matches.rs]: I feel that I've seen this pattern a few times now btw... maybe refactor in a follow up PR.

File: clippy_lints/src/matches.rs [hunk 4]
```diff
@@ -772,7 +754,7 @@ fn is_ref_some_arm(arm: &Arm) -> Option<BindingAnnotation> {
 fn has_only_ref_pats(arms: &[Arm]) -> bool {
     let mapped = arms
         .iter()
-        .flat_map(|a| &a.pats)
```
Comment (Centril) [inline clippy_lints/src/matches.rs]: I'd just inline into `arm.p.node` here.

File: clippy_lints/src/utils/mod.rs [hunk 5]
```diff
@@ -938,8 +938,8 @@ pub fn is_try(expr: &Expr) -> Option<&Expr> {
 
         if_chain! {
             if arms.len() == 2;
-            if arms[0].pats.len() == 1 && arms[0].guard.is_none();
-            if arms[1].pats.len() == 1 && arms[1].guard.is_none();
+            if arms[0].guard.is_none();
+            if arms[1].guard.is_none();
```
Comment (Centril) [inline clippy_lints/src/utils/mod.rs]: (Another instance of the same pattern)

File: clippy_lints/src/cognitive_complexity.rs (hunk 1 above)
Comment (Manishearth) [reply clippy_lints/src/cognitive_complexity.rs]: yeah i debated this but i think we might need a more careful CC accounting for pats here

**Comment by Manishearth** [comment]
Holding off on landing till https://github.com/rust-lang/rust/pull/64774 exists

**Comment by Manishearth** [comment]
Might be worth landing early via the merge button so we can pipeline up a clippyup PR though

**Comment by Manishearth** [comment]
@bors r=yaahc,centril

oops

**Comment by Manishearth** [comment]
@bors treeclosed- r=yaahc,centril

I'm marking OSX as an allowed failure so that we can get a clippyup landed immediately after the cargo update. There's three PRs worth of breakages here and I'd rather not buffer futher, especially since there's another big breakage coming soon.

**Pull request merged**